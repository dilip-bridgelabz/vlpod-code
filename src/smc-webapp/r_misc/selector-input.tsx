/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import * as React from "react";
import * as misc from "smc-util/misc";

const { FormControl, FormGroup } = require("react-bootstrap");

interface Props {
  options:
    | string[]
    | { value: string; display: JSX.Element | string }[]
    | { [keys: string]: JSX.Element }
    | Readonly<{ [keys: string]: string }>;
  disabled?: boolean;
  selected?: string;
  on_change?: (selected: string) => void;
  style?: React.CSSProperties;
}

// If the first element is a string, we assume the rest to be a string
function isStringArrayHeuristic(a: any): a is string[] {
  return typeof a[0] === "string";
}

export class SelectorInput extends React.Component<Props> {
  onChange = (e) => {
    if (this.props.on_change !== undefined) {
      this.props.on_change(e.target.value);
    }
  };

  render_options(): JSX.Element[] {
    const result: JSX.Element[] = [];
    if (Array.isArray(this.props.options)) {
      let x: any;
      if (isStringArrayHeuristic(this.props.options)) {
        let i = 0;
        for (x of this.props.options) {
          result.push(
            <option key={i} value={x}>
              {x}
            </option>
          );
          i += 1;
        }
        return result;
      } else {
        for (x of this.props.options) {
          result.push(
            <option key={x.value} value={x.value}>
              {x.display}
            </option>
          );
        }
        return result;
      }
    } else {
      const v = misc.keys(this.props.options);
      v.sort();
      for (const value of v) {
        const display = this.props.options[value];
        result.push(
          <option key={value} value={value}>
            {display}
          </option>
        );
      }
    }
    return result;
  }

  render() {
    return (
      <FormGroup style={this.props.style}>
        <FormControl
          value={this.props.selected}
          componentClass="select"
          onChange={this.onChange}
          disabled={this.props.disabled}
        >
          {this.render_options()}
        </FormControl>
      </FormGroup>
    );
  }
}
