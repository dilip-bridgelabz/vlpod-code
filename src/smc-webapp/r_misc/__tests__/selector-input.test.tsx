/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import * as React from "react";
import { shallow } from "enzyme";
import { SelectorInput } from "../selector-input";

describe("Smoke test:", () => {
  test("it renders a list of strings", () => {
    const names = ["Susan", "Harry", "Steve"];
    const tree = shallow(<SelectorInput options={names} />);
    expect(tree).toMatchSnapshot();
  });

  test("it renders a list of {value, display} objects", () => {
    const display_objects = [
      { value: "value_1", display: <div>Susan</div> },
      { value: "value_2", display: <div>Harry</div> },
    ];
    const tree = shallow(<SelectorInput options={display_objects} />)
      .children()
      .children();
    expect(tree).toMatchSnapshot();
  });

  test("it renders an object with keys mapped to react components ", () => {
    const options_map = {
      value_1: <div>Susan</div>,
      value_2: <div>Harry</div>,
    };
    const tree = shallow(<SelectorInput options={options_map} />)
      .children()
      .children();
    expect(tree).toMatchSnapshot();
  });
});

describe("Interactions:", () => {
  // Not the most robust testing, but propagation isn't well supported.
  // Might as well be explicit
  test("it calls on_change using e.target.value", () => {
    const names = ["Susan", "Harry", "Steve"];
    const selection_mock = jest.fn();

    const tree = shallow(
      <SelectorInput options={names} on_change={selection_mock} />
    );
    tree.children().simulate("change", { target: { value: "Susan" } });
    expect(selection_mock.mock.calls[0][0]).toBe("Susan");
  });
});
