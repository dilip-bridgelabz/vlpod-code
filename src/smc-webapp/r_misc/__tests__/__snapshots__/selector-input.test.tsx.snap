// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke test: it renders a list of {value, display} objects 1`] = `
Array [
  <option
    key="value_1"
    value="value_1"
  >
    <div>
      Susan
    </div>
  </option>,
  <option
    key="value_2"
    value="value_2"
  >
    <div>
      Harry
    </div>
  </option>,
]
`;

exports[`Smoke test: it renders a list of strings 1`] = `
<FormGroup
  bsClass="form-group"
>
  <FormControl
    bsClass="form-control"
    componentClass="select"
    onChange={[Function]}
  >
    <option
      key="0"
      value="Susan"
    >
      Susan
    </option>
    <option
      key="1"
      value="Harry"
    >
      Harry
    </option>
    <option
      key="2"
      value="Steve"
    >
      Steve
    </option>
  </FormControl>
</FormGroup>
`;

exports[`Smoke test: it renders an object with keys mapped to react components  1`] = `
Array [
  <option
    key="value_1"
    value="value_1"
  >
    <div>
      Susan
    </div>
  </option>,
  <option
    key="value_2"
    value="value_2"
  >
    <div>
      Harry
    </div>
  </option>,
]
`;
